File: src/lib/demo-user.ts
import { cookies } from "next/headers";
import type { User } from "@prisma/client";
import { prisma } from "./db";

const COOKIE_NAME = "fin_demo_user_id";
const COOKIE_MAX_AGE_SECONDS = 60 * 60 * 24 * 365;

/**
 * Single auto-provisioned "demo user" per browser, identified by a
 * long-lived httpOnly cookie — no login UI (see design spec, Section 1).
 *
 * Server Components can read cookies but not set them (Next.js
 * restriction: only Server Actions / Route Handlers may call
 * `cookies().set()`). Called from the dashboard (a Server Component)
 * with no cookie yet, this still creates the row (needed to query, even
 * though it'll have no data) but silently can't persist the cookie — the
 * dashboard immediately redirects to /onboarding in that case anyway.
 * The wizard's submit route (a Route Handler) is where the cookie
 * actually gets set for real.
 */
export async function getOrCreateDemoUser(): Promise<User> {
  const cookieStore = await cookies();
  const existingId = cookieStore.get(COOKIE_NAME)?.value;

  if (existingId) {
    const existing = await prisma.user.findUnique({ where: { id: existingId } });
    if (existing) return existing;
  }

  const created = await prisma.user.create({
    data: { email: `demo-${crypto.randomUUID()}@fin.local` },
  });

  try {
    cookieStore.set(COOKIE_NAME, created.id, {
      httpOnly: true,
      maxAge: COOKIE_MAX_AGE_SECONDS,
      sameSite: "lax",
      path: "/",
    });
  } catch {
    // Not callable from a Server Component — see doc comment above.
  }

  return created;
}

File: backend/app/demo_user.py
import os
import uuid

from fastapi import Depends, Request, Response
from sqlalchemy.orm import Session

from app.db import get_db
from app.models import User

COOKIE_NAME = "fin_demo_user_id"
COOKIE_MAX_AGE_SECONDS = 60 * 60 * 24 * 365


def get_or_create_demo_user(
    request: Request,
    response: Response,
    db: Session = Depends(get_db),
) -> User:
    existing_id = request.cookies.get(COOKIE_NAME)

    if existing_id:
        existing = db.get(User, existing_id)
        if existing:
            return existing

    created = User(email=f"demo-{uuid.uuid4()}@fin.local")
    db.add(created)
    db.commit()
    db.refresh(created)

    is_production = os.getenv("ENVIRONMENT", "development") == "production"
    response.set_cookie(
        key=COOKIE_NAME,
        value=created.id,
        max_age=COOKIE_MAX_AGE_SECONDS,
        httponly=True,
        samesite="none" if is_production else "lax",
        secure=is_production,
        path="/",
    )

    return created
